### PR TITLE
Fix Jenkins API integration

### LINKED ISSUE #101: Can't perform REST queries if AzureAD user doesn't exist explicitly in Auth Matrix
### Version report

Jenkins and plugins versions report:

```
Jenkins: 2.263.2
AzureAD plugin: 1.2.1

```

- What Operating System are you using (both controller, and any agents involved in the problem)?

```
Centos 7
```

### Reproduction steps
My user has been added to the AzureAD group, this group was mapped with admin permissions to Jenkins.
The user doesn't exist explicitly in Jenkins Auth Matrix.

query1:
```
curl -XPOST -u "me@company.com:TOKEN"  https://JENKINS_URL/job/devops/job/docker/build
```

result1:
```
<h1>Access Denied</h1><p class="error">me@company.com is missing the Overall/Read permission</div>
```

At the same time, I have a service account (non-real user), **which has been added explicitly to the Jenkins Auth Matrix.**
If I perform a query - everything works as expected.  

query2:
```
curl -XPOST -u "sevice_account@company.com:TOKEN"  https://JENKINS_URL/job/devops/job/docker/build
```

result2:
```
  (no answer - that is 200 OK)
```

I figure out that I have to add all AzureAD group's users explicitly to the Jenkins Auth Matrix to have for all of the AzureAD users capabilities to run jobs via REST API. 
Seems the AzureAD plugin knows nothing about the particular user when it **has no entry in Jenkins Auth Matrix** and you trying to trigger REST.

Thanks in advance,
Pasha.

### LINKED ISSUE #105: CLI is not working for AD groups set in Role-based authorization
<p>We use Azure AD authentication with Role-based authorization plugin, it works perfectly from the web pages. In the authorization options we use email addresses for users and group names for groups, both of these works.</p>

<p>However, when we try to use the CLI, the groups are not found. In the attached log, the user has the Overall/Read permission through a group, but that is not found. If I give the permission directly for a single user (without AD groups), it works.</p>

<p>Interestingly, when I use direct permissions, the error message "The user *** may or may not exist in the SecurityRealm, so we provide minimum access" still appears in the logs, but the CLI itself is working.</p>

<p> </p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">
Dec 15 10:29:49 jenkins tomcat9[1599]: Authentication attempted from email@redacted.com with Sun RSA <span class="code-keyword">public</span> key, 2048 bits
Dec 15 10:29:49 jenkins tomcat9[1599]:   params: <span class="code-keyword">null</span>
Dec 15 10:29:49 jenkins tomcat9[1599]:   modulus: *****
Dec 15 10:29:49 jenkins tomcat9[1599]:   <span class="code-keyword">public</span> exponent: 65537
Dec 15 10:29:49 jenkins tomcat9[1599]: The user email@redacted.com may or may not exist in the SecurityRealm, so we provide minimum access
Dec 15 10:29:49 jenkins tomcat9[1599]: authenticated: email@redacted.com []
Dec 15 10:29:49 jenkins tomcat9[1599]: doAuth(email@redacted.com@ServerSessionImpl[<span class="code-keyword">null</span>@/10.0.100.9:52238]) key type=ssh-rsa, fingerprint=SHA256:***** - authentication result: <span class="code-keyword">true</span>
Dec 15 10:29:49 jenkins tomcat9[1599]: doAuth(email@redacted.com@ServerSessionImpl[<span class="code-keyword">null</span>@/10.0.100.9:52238]) send SSH_MSG_USERAUTH_PK_OK <span class="code-keyword">for</span> key type=ssh-rsa, fingerprint=SHA256:*****
Dec 15 10:29:49 jenkins tomcat9[1599]: writePacket(Nio2Session[local=/10.0.210.1:2222, remote=/10.0.100.9:52238]) Writing 352 bytes
Dec 15 10:29:49 jenkins tomcat9[1599]: handleCompletedWriteCycle(Nio2Session[local=/10.0.210.1:2222, remote=/10.0.100.9:52238]) finished writing len=352
Dec 15 10:29:49 jenkins tomcat9[1599]: handleAuthenticationInProgress(email@redacted.com@ServerSessionImpl[<span class="code-keyword">null</span>@/10.0.100.9:52238]) SSH_MSG_USERAUTH_REQUEST
Dec 15 10:29:49 jenkins tomcat9[1599]: handleReadCycleCompletion(Nio2Session[local=/10.0.210.1:2222, remote=/10.0.100.9:52238]) read 688 bytes
Dec 15 10:29:49 jenkins tomcat9[1599]: process(ServerSessionImpl[<span class="code-keyword">null</span>@/10.0.100.9:52238]) Received SSH_MSG_USERAUTH_REQUEST user=email@redacted.com, service=ssh-connection, method=publickey
Dec 15 10:29:49 jenkins tomcat9[1599]: process(ServerSessionImpl[<span class="code-keyword">null</span>@/10.0.100.9:52238]) Authenticating user <span class="code-quote">'email@redacted.com'</span> with service <span class="code-quote">'ssh-connection'</span> and method <span class="code-quote">'publickey'</span> (attempt 2 / 20)
Dec 15 10:29:49 jenkins tomcat9[1599]: doAuth(email@redacted.com@ServerSessionImpl[<span class="code-keyword">null</span>@/10.0.100.9:52238]) verify key type=ssh-rsa, factories=ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,ssh-ed25519,ssh-rsa,ssh-dss, fingerprint=SHA256:*****
Dec 15 10:29:49 jenkins tomcat9[1599]: Authentication attempted from email@redacted.com with Sun RSA <span class="code-keyword">public</span> key, 2048 bits
Dec 15 10:29:49 jenkins tomcat9[1599]:   params: <span class="code-keyword">null</span>
Dec 15 10:29:49 jenkins tomcat9[1599]:   modulus: *****
Dec 15 10:29:49 jenkins tomcat9[1599]:   <span class="code-keyword">public</span> exponent: 65537
Dec 15 10:29:49 jenkins tomcat9[1599]: The user email@redacted.com may or may not exist in the SecurityRealm, so we provide minimum access
Dec 15 10:29:49 jenkins tomcat9[1599]: authenticated: email@redacted.com []
Dec 15 10:29:49 jenkins tomcat9[1599]: doAuth(email@redacted.com@ServerSessionImpl[<span class="code-keyword">null</span>@/10.0.100.9:52238]) key type=ssh-rsa, fingerprint=SHA256:***** - authentication result: <span class="code-keyword">true</span>
Dec 15 10:29:49 jenkins tomcat9[1599]: doAuth(email@redacted.com@ServerSessionImpl[<span class="code-keyword">null</span>@/10.0.100.9:52238]) key type=ssh-rsa, fingerprint=SHA256:***** - verified
Dec 15 10:29:49 jenkins tomcat9[1599]: handleAuthenticationSuccess(email@redacted.com@ServerSessionImpl[<span class="code-keyword">null</span>@/10.0.100.9:52238]) SSH_MSG_USERAUTH_REQUEST
Dec 15 10:29:49 jenkins tomcat9[1599]: writePacket(Nio2Session[local=/10.0.210.1:2222, remote=/10.0.100.9:52238]) Writing 64 bytes
Dec 15 10:29:49 jenkins tomcat9[1599]: handleCompletedWriteCycle(Nio2Session[local=/10.0.210.1:2222, remote=/10.0.100.9:52238]) finished writing len=64
Dec 15 10:29:49 jenkins tomcat9[1599]: Session email@redacted.com@/10.0.100.9:52238 authenticated
Dec 15 10:29:49 jenkins tomcat9[1599]: handleReadCycleCompletion(Nio2Session[local=/10.0.210.1:2222, remote=/10.0.100.9:52238]) read 80 bytes
Dec 15 10:29:49 jenkins tomcat9[1599]: channelOpen(ServerConnectionService[ServerSessionImpl[email@redacted.com@/10.0.100.9:52238]]) SSH_MSG_CHANNEL_OPEN sender=0, type=session, window-size=2097152, packet-size=32768
Dec 15 10:29:49 jenkins tomcat9[1599]: init() service=ServerConnectionService[ServerSessionImpl[email@redacted.com@/10.0.100.9:52238]] session=ServerSessionImpl[email@redacted.com@/10.0.100.9:52238] id=0
Dec 15 10:29:49 jenkins tomcat9[1599]: init(Window[server/local](ChannelSession[id=0, recipient=-1]-ServerSessionImpl[email@redacted.com@/10.0.100.9:52238])) size=2097152, max=2097152, packet=32768
Dec 15 10:29:49 jenkins tomcat9[1599]: registerChannel(ServerConnectionService[ServerSessionImpl[email@redacted.com@/10.0.100.9:52238]])[id=0] ChannelSession[id=0, recipient=-1]-ServerSessionImpl[email@redacted.com@/10.0.100.9:52238]
Dec 15 10:29:49 jenkins tomcat9[1599]: setRecipient(ChannelSession[id=0, recipient=-1]-ServerSessionImpl[email@redacted.com@/10.0.100.9:52238]) recipient=0
Dec 15 10:29:49 jenkins tomcat9[1599]: init(Window[server/remote](ChannelSession[id=0, recipient=0]-ServerSessionImpl[email@redacted.com@/10.0.100.9:52238])) size=2097152, max=2097152, packet=32768
Dec 15 10:29:49 jenkins tomcat9[1599]: init(Window[server/local](ChannelSession[id=0, recipient=0]-ServerSessionImpl[email@redacted.com@/10.0.100.9:52238])) re-initializing
Dec 15 10:29:49 jenkins tomcat9[1599]: init(Window[server/local](ChannelSession[id=0, recipient=0]-ServerSessionImpl[email@redacted.com@/10.0.100.9:52238])) size=2097152, max=2097152, packet=32768
Dec 15 10:29:49 jenkins tomcat9[1599]: operationComplete(ChannelSession[id=0, recipient=0]-ServerSessionImpl[email@redacted.com@/10.0.100.9:52238]) send SSH_MSG_CHANNEL_OPEN_CONFIRMATION recipient=0, sender=0, window-size=2097152, packet-size=32768
Dec 15 10:29:49 jenkins tomcat9[1599]: writePacket(Nio2Session[local=/10.0.210.1:2222, remote=/10.0.100.9:52238]) Writing 80 bytes
Dec 15 10:29:49 jenkins tomcat9[1599]: handleCompletedWriteCycle(Nio2Session[local=/10.0.210.1:2222, remote=/10.0.100.9:52238]) finished writing len=80
Dec 15 10:29:49 jenkins tomcat9[1599]: handleReadCycleCompletion(Nio2Session[local=/10.0.210.1:2222, remote=/10.0.100.9:52238]) read 112 bytes
Dec 15 10:29:49 jenkins tomcat9[1599]: handleChannelRequest(ChannelSession[id=0, recipient=0]-ServerSessionImpl[email@redacted.com@/10.0.100.9:52238]) SSH_MSG_CHANNEL_REQUEST exec wantReply=<span class="code-keyword">true</span>
Dec 15 10:29:49 jenkins tomcat9[1599]: handleExec(ChannelSession[id=0, recipient=0]-ServerSessionImpl[email@redacted.com@/10.0.100.9:52238]) Executing command: console Core-Gen-Master/Core-Run-Test 1419
Dec 15 10:29:49 jenkins tomcat9[1599]: prepareChannelCommand(ChannelSession[id=0, recipient=0]-ServerSessionImpl[email@redacted.com@/10.0.100.9:52238])[exec] prepared command
Dec 15 10:29:49 jenkins tomcat9[1599]: sendResponse(ChannelSession[id=0, recipient=0]-ServerSessionImpl[email@redacted.com@/10.0.100.9:52238]) request=exec result=ReplySuccess, want-reply=<span class="code-keyword">true</span>
Dec 15 10:29:49 jenkins tomcat9[1599]: writePacket(Nio2Session[local=/10.0.210.1:2222, remote=/10.0.100.9:52238]) Writing 64 bytes
Dec 15 10:29:49 jenkins tomcat9[1599]: handleCompletedWriteCycle(Nio2Session[local=/10.0.210.1:2222, remote=/10.0.100.9:52238]) finished writing len=64
Dec 15 10:29:49 jenkins tomcat9[1599]: sendResponse(ChannelSession[id=0, recipient=0]-ServerSessionImpl[email@redacted.com@/10.0.100.9:52238]) request=exec activate command
Dec 15 10:29:49 jenkins tomcat9[1599]: The user email@redacted.com may or may not exist in the SecurityRealm, so we provide minimum access
Dec 15 10:29:49 jenkins tomcat9[1599]: The user email@redacted.com may or may not exist in the SecurityRealm, so we provide minimum access
Dec 15 10:29:49 jenkins tomcat9[1599]: hasPermission(org.acegisecurity.providers.UsernamePasswordAuthenticationToken@2d87e98f: Username: email@redacted.com; Password: [PROTECTED]; Authenticated: <span class="code-keyword">true</span>; Details: <span class="code-keyword">null</span>; Granted Authorities: authenticated,Permission[<span class="code-keyword">class </span>hudson.model.Hudson,Read])=><span class="code-keyword">null</span>, thus <span class="code-keyword">false</span>
Dec 15 10:29:49 jenkins tomcat9[1599]: Failed call to CLI command console, with 2 arguments, as user email@redacted.com.
Dec 15 10:29:49 jenkins tomcat9[1599]: hudson.security.AccessDeniedException2: email@redacted.com is missing the Overall/Read permission
Dec 15 10:29:49 jenkins tomcat9[1599]:     at hudson.security.ACL.checkPermission(ACL.java:79)
Dec 15 10:29:49 jenkins tomcat9[1599]:     at hudson.security.AccessControlled.checkPermission(AccessControlled.java:47)
Dec 15 10:29:49 jenkins tomcat9[1599]:     at hudson.cli.CLICommand.main(CLICommand.java:245)
Dec 15 10:29:49 jenkins tomcat9[1599]:     at org.jenkinsci.main.modules.sshd.CLICommandAdapter$1.run(CLICommandAdapter.java:37)
Dec 15 10:29:49 jenkins tomcat9[1599]:     at org.jenkinsci.main.modules.sshd.AsynchronousCommand$1.run(AsynchronousCommand.java:112)
Dec 15 10:29:49 jenkins tomcat9[1599]:     at java.base/java.lang.<span class="code-object">Thread</span>.run(<span class="code-object">Thread</span>.java:834) </pre>
</div></div>
<p> </p>
<i>[JENKINS-64447] created by ngg1</i>

### PR DESCRIPTION
**Pull request by hikkyXII**
AzureSecurityRealm overrides **loadUserByUsername(String username)** function from **UserDetailsService**.
Somewhere Jenkins is using this function in permissions check, for example when triggering one build from another or like described in #101.

Fixes #101
Fixes #105

I am not spec in Java or Azure so couldn't implement request for UserDetails from Azure in **loadUserByUsername**. 
But I could created Singleton cache class that keeps user details and gives it to Jenkins.

It solves missing permissions issues. I really require this fix in my instances. 

@timja , @xuzhang3, as I understand you are most active contributors, could you please review/comment/approve.

### CODE REVIEW

**Comment by timja** [comment]
Planning to look at this _soon-ish_, I don't think this is the correct fix as from a quick look I think it will only handle users who have logged in via the web-ui in the last 4 hours.

I think the code needs to be updated to query AzureAD (or user the `id_token` if supplied) and the id_token could be an alternative to the api token of Jenkins

From my initial look I saw a lot of deprecated APIs so have cleaned it up in https://github.com/jenkinsci/azure-ad-plugin/pull/113

**Comment by hikkyXII** [comment]
#113  Massive clean...

Agree, it is workaround not fix and for API calls works only for cache duration time after user login.

Suppose unique name(PrincipalName) would be better if it is possible to use. It is human readable, useful for logs.

File: src/main/java/com/microsoft/jenkins/azuread/AzureAdUser.java [hunk 1]
```diff
@@ -52,12 +53,31 @@ private AzureAdUser() {
         authorities = Arrays.asList(SecurityRealm.AUTHENTICATED_AUTHORITY2);
     }
 
+
+
```
Comment (timja) [inline src/main/java/com/microsoft/jenkins/azuread/AzureAdUser.java:57]:
```suggestion
```

**Comment by timja** [comment]
@hikkyXII does this meet your requirements?

**Comment by hikkyXII** [comment]
@timja yes, looks cool and clean.
But I see exceptions in **Configure Global Security** ->**Authorization** -> **Azure Active Directory Matrix-based security**
User/group from Azure are colored in red and **UserMayOrMayNotExistException2** exception output.

**Comment by timja** [comment]
Is there anything in your logs? there's a few different ways to configure this plugin I believe.

**Comment by hikkyXII** [comment]
No, nothing in Jenkins output.
What are different ways?

**Comment by timja** [comment]
Have you granted permissions for the service principal to read directory data? I did that via adding a role assignment for read directory data but you can also do it with api permissions I believe and some people use id_token only for auth and don’t grant any extra permissions

**Comment by hikkyXII** [comment]
Yes, permissions are granted, I can search for users with autocomplete. 
In **Azure Active Directory Matrix-based security** names are like **_\<display Name> (\<OID>)_** and it now works for **azureClient.activeDirectoryUsers().getByName(username)** and **UserMayOrMayNotExistException2** is not caught properly somewhere higher.

**Comment by timja** [comment]
Right I see where you mean, groups aren't being handled properly in the new code just usernames

**Comment by hikkyXII** [comment]
Not only groups, single users too.

**Comment by timja** [comment]
> Not only groups, single users too.

Should be fine now, groups were still broken with your change

**Comment by hikkyXII** [comment]
Cool.
@timja, Thanks a lot!

**Pull request merged**